### PR TITLE
common, utility, rxd, sockets: Fix warnings produced by MSVC on Windows

### PR DESCRIPTION
**Pull request by gladkovdmitry17**

### CODE REVIEW

File: include/windows/osd.h [hunk 1]
```diff
@@ -694,6 +694,12 @@ static inline SOCKET ofi_socket(int domain, int type, int protocol)
 	return socket(domain, type, protocol);
 }
 
+static inline int ofi_setsockopt(SOCKET fd, int level, int optname,
```
Comment (gladkovdmitry17) [inline include/windows/osd.h]: @shefty It was introduced to avoid conversation from any types to (const char *) optval on Windows.

File: prov/sockets/include/sock.h [hunk 2]
```diff
@@ -413,13 +413,13 @@ enum {
...
-		uint8_t	datatype;
-		uint8_t	res_iov_len;
-		uint8_t	cmp_iov_len;
+		uint8_t		op;
+		uint8_t		datatype;
+		uint64_t	res_iov_len;
+		uint64_t	cmp_iov_len;
 	} atomic;
 	uint8_t	reserved[1];
 };
```
Comment (shefty) [inline prov/sockets/include/sock.h]: This structure is part of the wire protocol.  We need to avoid changing it.

File: prov/sockets/src/sock_cntr.c [hunk 3]
```diff
@@ -318,7 +319,7 @@ static int sock_cntr_wait(struct fid_cntr *fid_cntr, uint64_t threshold,
 		goto out;
 	}
 
-	if (ofi_atomic_get32(&cntr->value) >= (int)threshold) {
+	if (ofi_atomic_get64(&cntr->value) >= (int)threshold) {
```
Comment (shefty) [inline prov/sockets/src/sock_cntr.c]: change threshold cast from int to int64_t

File: prov/sockets/src/sock_cntr.c [hunk 4]
```diff
@@ -512,12 +513,11 @@ int sock_cntr_open(struct fid_domain *domain, struct fi_cntr_attr *attr,
 		goto err;
 
 	if (attr == NULL)
-		memcpy(&_cntr->attr, &sock_cntr_add, sizeof(sock_cntr_attr));
+		memcpy(&_cntr->attr, &sock_cntr_attr, sizeof(sock_cntr_attr));
```
Comment (shefty) [inline prov/sockets/src/sock_cntr.c]: Can you separate this into its own patch?  This looks like a bug fix.

File: prov/sockets/src/sock_wait.c [hunk 5]
```diff
@@ -75,8 +75,9 @@ int sock_wait_get_obj(struct fid_wait *fid, void *arg)
 		SOCK_LOG_ERROR("Invalid wait obj type\n");
 		return -FI_EINVAL;
 	}
-
+#ifndef _WIN32
 	return 0;
+#endif
```
Comment (shefty) [inline prov/sockets/src/sock_wait.c]: This looks wrong for windows.  Shouldn't we return _something_ here?  ENOSYS?

**Comment by shefty** [comment]
See build error:
prov/rxd/src/rxd_ep.c:531:18: error: format specifies type 'int' but the argument has type 'ssize_t' (aka 'long') [-Werror,-Wformat]
                        ctrl->seg_no, ret);

File: prov/sockets/src/sock_cntr.c (hunk 4 above)
Comment (gladkovdmitry17) [reply prov/sockets/src/sock_cntr.c]: Sure, I can.
Btw, there are multiple places where a function is declared to return `size_t`, but there are branches in the function that returns `-1`. Should I separate them as well?

File: prov/sockets/include/sock.h (hunk 2 above)
Comment (gladkovdmitry17) [reply prov/sockets/include/sock.h]: @shefty , sure. there are a couple of similiar to that places. I'll fix all.
Should it be fixed for the `struct sock_msg_hdr`?

Comment (shefty) [reply prov/sockets/include/sock.h]: We don't want to modify the wire protocol, so casts will be needed when filling out the wire structure fields.

**Comment by lanl-ompi** [comment]
Can one of the admins verify this patch?

**Comment by hppritcha** [comment]
add to whitelist

**Comment by shefty** [comment]
See PR #3625.  This PR is out of date with several conflicts.  New PR has been open with changes going there to fix issues.  Until 3625 has been merged, please keep this PR open for reference.  (Yes, I finally decided to setup my system to build for windows.)

**Comment by gladkovdmitry17** [comment]
> Yes, I finally decided to setup my system to build for windows.

Unfortunately, I have no time to put necessary effort for this PR at the moment
Thanks, @shefty. I can support you and review the prepared #3625 PR
